cli/src/commands/parse.rs

```diff
@@ -121,14 +128,91 @@ pub fn parse_lockfile(
     let packages = parse_lockfile_content(&generated_lockfile, parser)?;
 
     Ok(ParsedLockfile::new(display_path, format, packages))
 }
 
+/// Generate a lockfile from a manifest inside a sandbox.
+fn generate_lockfile(generator: &dyn Generator, path: &Path, sandbox: bool) -> Result<String> {
+    let canonical_path = path.canonicalize()?;
+
+    // Enable the sandbox.
+    if sandbox {
+        let birdcage = lockfile_generation_sandbox(&canonical_path)?;
+        birdcage.lock()?;
+    }
+
+    let generated_lockfile = generator
+        .generate_lockfile(&canonical_path)
+        .context("Lockfile generation failed! For details, see: \
+            https://docs.phylum.io/docs/lockfile_generation")?;
+
+    Ok(generated_lockfile)
+}
+
+/// Create sandbox with exceptions allowing generation of any lockfile.
+fn lockfile_generation_sandbox(canonical_manifest_path: &Path) -> Result<Birdcage> {
+    let mut birdcage = permissions::default_sandbox()?;
+
+    // Allow all networking.
+    birdcage.add_exception(Exception::Networking)?;
+
+    // Add exception for the manifest's parent directory.
+    let project_path = canonical_manifest_path.parent().expect("Invalid manifest path");
+    permissions::add_exception(&mut birdcage, Exception::WriteAndRead(project_path.into()))?;
+
+    // Add exception for all the executables required for generation.
+    let ecosystem_bins = [
+        "cargo", "bundle", "mvn", "gradle", "npm", "pnpm", "yarn", "python3", "pipenv", "poetry",
+        "go", "dotnet",
+    ];
+    for bin in ecosystem_bins {
+        let absolute_path = permissions::resolve_bin_path(bin);
+        permissions::add_exception(&mut birdcage, Exception::ExecuteAndRead(absolute_path))?;
+    }
+
+    // Allow any executable in common binary directories.
+    //
+    // Reading binaries shouldn't be an attack vector, but significantly simplifies
+    // complex ecosystems (like Python's symlinks).
+    permissions::add_exception(&mut birdcage, Exception::ExecuteAndRead("/usr/bin".into()))?;
+    permissions::add_exception(&mut birdcage, Exception::ExecuteAndRead("/bin".into()))?;
+
+    // Add paths required by specific ecosystems.
+    //
+    // This will automatically resolve `~` to the user's home directory.
+    #[allow(clippy::type_complexity)]
+    let path_exceptions: &[(_, fn(PathBuf) -> Exception)] = &[
+        // Cargo.
+        ("~/.rustup", Exception::ExecuteAndRead),
+        ("~/.cargo", Exception::Read),
+        ("/etc/passwd", Exception::Read),
+        // Bundle.
+        ("/dev/urandom", Exception::Read),
+        // Maven
+        ("/opt/maven", Exception::ExecuteAndRead),
+        ("/etc/java-openjdk", Exception::Read),
+        // Gradle.
+        ("/usr/share/java/gradle/lib", Exception::Read),
+        // Pnpm.
+        ("/tmp", Exception::Read),
+        // Yarn.
+        ("~/.yarn", Exception::Read),
+    ];
+    let home = dirs::home_dir()?;
+    let home = home.to_string_lossy();
```

Comment:
I don't like requiring a lossy conversion to UTF-8 here. I know invalid UTF-8 in a path (especially the home path) is unlikely... But `home` is a `PathBuf` and the end result we need is a `PathBuf`, so converting to a `String` just for `.replace()` feels kinda wrong.

One alternative would be to use relative paths for anything in the home directory and then use `home.join(path)`. If `path` is absolute, that will replace `home` entirely thanks to how `.join()` works. But maybe that's tricky to read...

Honestly, I wonder if this code would be cleaner if it just called `permissions::add_exception()` 9 times instead of using the loop. That would also remove the need for the `#[allow(clippy::type_complexity)]`.

Reply:
I've unrolled the loop, let me know if you like that more.

I'd also be curious if these Java paths work on macOS.